hivemind/dht/node.py

```diff
@@ -128,145 +130,244 @@ async def shutdown(self, timeout=None):
+            num_workers: Optional[int] = None, beam_size: Optional[int] = None
+    ) -> Dict[DHTKey, Tuple[Optional[DHTValue], Optional[DHTExpiration]]]:
+        """
+        :param keys: traverse the DHT and find the value for each of these keys (or (None, None) if not key found)
         :param sufficient_expiration_time: if the search finds a value that expires after this time,
             default = time of call, find any value that did not expire by the time of call
             If min_expiration_time=float('inf'), this method will find a value with _latest_ expiration
         :param beam_size: maintains up to this many nearest nodes when crawling dht, default beam_size = bucket_size
-        :returns: value and its expiration time. If nothing is found , returns (None, None).
+        :param num_workers: override for default num_workers, see traverse_dht num_workers param
+        :returns: for each key: value and its expiration time. If nothing is found , returns (None, None) for that key
         :note: in order to check if get returned a value, please check (expiration_time is None)
         """
-        key_id = DHTID.generate(key)
+        key_ids = [DHTID.generate(key) for key in keys]
+        id_to_original_key = dict(zip(key_ids, keys))
         sufficient_expiration_time = sufficient_expiration_time or get_dht_time()
         beam_size = beam_size if beam_size is not None else self.protocol.bucket_size
-        latest_value_bytes, latest_expiration, latest_node_id = b'', -float('inf'), None
-        node_to_addr, nodes_checked_for_value, nearest_nodes = dict(), set(), []
-        should_cache = False  # True if found value in DHT that is newer than local value
-
-        # Option A: value can be stored in our local cache
-        maybe_value, maybe_expiration = self.protocol.storage.get(key_id)
-        if maybe_expiration is None:
-            maybe_value, maybe_expiration = self.protocol.cache.get(key_id)
-        if maybe_expiration is not None and maybe_expiration > latest_expiration:
-            latest_value_bytes, latest_expiration, latest_node_id = maybe_value, maybe_expiration, self.node_id
-            # TODO(jheuristic) we may want to run background beam search to update our cache
-        nodes_checked_for_value.add(self.node_id)
-
-        # Option B: go beam search the DHT
-        if latest_expiration < sufficient_expiration_time:
+        num_workers = num_workers if num_workers is not None else self.num_workers
+
+        # search metadata
+        unfinished_key_ids = set(key_ids)  # track key ids for which the search is not terminated
+        node_to_addr: Dict[DHTID, Endpoint] = dict()  # global routing table for all queries
+        latest_result = {key_id: (b'', -float('inf'), None) for key_id in key_ids}  # key: (value_bin, expiration, node)
+        BINARY_VALUE, EXPIRATION, NODE = 0, 1, 2  # values in latest_result (for readability)
```

Comment:
Enum?

Reply:
would you tolerate a namedtuple?

Reply:
upd: implemented a namedtuple
![image](https://user-images.githubusercontent.com/3491902/86533007-bb20f700-bed6-11ea-9787-2a4adebfca3c.png)